### PR TITLE
Architecture: Replace Read/Write Substate interface with SubstateRef and SubstateRefMut

### PR DESCRIPTION
**Pull request by talekhinezh**
* Replace substate read/write semantics with SubstateRef and SubstateRefMut
* Moved substate read/write semantics into WASM Runtime
* Use SubstateRef interface in some native blueprints (`resource_manager` and `component`)

### CODE REVIEW

File: radix-engine/src/model/substates/substate.rs [hunk 1]
```diff
@@ -266,3 +261,337 @@ impl Into<GlobalAddressSubstate> for Substate {
...
+                Substate::KeyValueStoreEntry(next),
+            ) => *current = next,
+            (RawSubstateRefMut::NonFungible(current), Substate::NonFungible(next)) => {
+                *current = next
+            }
+            (RawSubstateRefMut::System(current), Substate::System(next)) => *current = next,
+            _ => return Err(RuntimeError::KernelError(KernelError::InvalidOverwrite)),
+        }
+
+        self.flush()
```
Comment (iamyulong) [inline radix-engine/src/model/substates/substate.rs:500]: Better not to flush automatically?

File: radix-engine/src/model/substates/substate.rs [hunk 2]
```diff
@@ -266,3 +261,337 @@ impl Into<GlobalAddressSubstate> for Substate {
...
+    }
+
+    pub fn get_raw_mut(&mut self) -> RawSubstateRefMut {
```
Comment (iamyulong) [inline radix-engine/src/model/substates/substate.rs:503]: Why do we need two write interfaces, `overwrite` and `get_raw_mut`? With a raw mutable reference, we can already overwrite the value.

File: radix-engine/src/model/substates/substate.rs [hunk 3]
```diff
@@ -266,3 +261,337 @@ impl Into<GlobalAddressSubstate> for Substate {
...
+            track,
+        };
+        Ok(substate_ref_mut)
+    }
+
+    pub fn offset(&self) -> &SubstateOffset {
+        &self.offset
+    }
+
+    pub fn flush(&mut self) -> Result<(), RuntimeError> {
```
Comment (iamyulong) [inline radix-engine/src/model/substates/substate.rs:443]: Should we consume the `RefMut` upon `flush`? It seems flush can only be done once.

File: radix-engine/src/engine/modules/module.rs [hunk 4]
```diff
@@ -31,9 +31,8 @@ pub enum SysCallInput<'a> {
     ReadSubstate {
         lock_handle: &'a LockHandle,
     },
-    WriteSubstate {
+    GetMut {
```
Comment (iamyulong) [inline radix-engine/src/engine/modules/module.rs:34]: Minor: Can we rename to `GetRefMut`/`get_ref_mut` for consistence?

File: radix-engine/src/model/substates/substate.rs (hunk 1 above)
Comment (talekhinezh) [reply radix-engine/src/model/substates/substate.rs:500]: This was really a convenience method to "flash" a substate. Will see if it makes sense to remove this.

File: radix-engine/src/model/substates/substate.rs (hunk 2 above)
Comment (talekhinezh) [reply radix-engine/src/model/substates/substate.rs:503]: `overwrite` was really a convenience method to "flash" a substate. Will see if it makes sense to remove this.

File: radix-engine/src/model/substates/substate.rs (hunk 3 above)
Comment (talekhinezh) [reply radix-engine/src/model/substates/substate.rs:443]: hrm, good idea! flush could be called multiple times currently but probably simpler to only allow it once.

**Review by iamyulong — APPROVED** [review]

**Pull request merged**